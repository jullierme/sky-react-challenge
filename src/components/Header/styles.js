import styled from 'styled-components';
import { Link } from 'react-router-dom';

export const LinkMenu = styled(Link)`
  font-size: 1.3rem;
  font-weight: 400;
  color: #fff;
  padding-right: 1em;
`;

export const Container = styled.div`
  height: 80px;
  padding: 0 1em;
  background: #37003c;
  color: #fff;
  display: flex;
  align-items: center;
`;
