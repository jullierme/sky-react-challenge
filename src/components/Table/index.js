import React, { Component } from 'react';

import { fetch } from '../../services/api';

class Table extends Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  // TASK #4 - create a table of results
  computeTable(teams, weeksMatches) {
    return [];
  }

  componentWillMount() {
    Promise.all([fetch('/teams'), fetch('/weeks')]).then(
      (teams, weeksMatches) => {
        this.setState({
          table: this.computeTable(teams, weeksMatches),
        });
      }
    );
  }

  render() {
    const t = this.state.table;
    if (!t) return <div>loading...</div>;

    return 'TODO 4';
  }
}

export default Table;
