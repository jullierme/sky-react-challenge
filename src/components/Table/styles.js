import styled from 'styled-components';

export const TableOfResults = styled.table`
  width: 100%;
  text-align: center;
  font-size: 1rem;
`;

export const Thead = styled.thead`
  background-color: #fbfbfb;

  th {
    color: #6c6c6c;
    font-size: 1rem;
    padding: 10px 0;
  }
`;

export const Tbody = styled.tbody``;

export const TdTeam = styled.td`
  text-align: left;

  img {
    height: 30px;
    margin-right: 10px;
  }

  a {
    display: flex;
    align-items: center;
  }
`;
