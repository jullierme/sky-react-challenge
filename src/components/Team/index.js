import React, { useEffect, useState } from 'react';
import Results from '../Results';
import { fetch } from '../../services/api';
import { getComputedTable } from '../../services/stats';

import {
  Body,
  Header,
  Container,
  StatsContainer,
  ResultsContainer,
  Label,
  Stat,
} from './styles';

function Team({ match }) {
  const [team, setTeam] = useState([]);
  const [loading, setLoading] = useState(false);
  const [stats, setStats] = useState({});

  useEffect(() => {
    fetchTeam();
  }, [match.params.index]);

  const fetchTeam = async () => {
    setLoading(true);

    const data = await fetch(`/teams/${match.params.index}`);
    data.id = parseInt(data.id, 10);

    setTeam(data);
    setStats(getTeamStats(data.id, data.results));

    setLoading(false);
  };

  // TASK #3 - compute team stats
  const getTeamStats = (id, results) => {
    const table = getComputedTable([results]);

    return table.find(club => {
      return club.id === id;
    });
  };

  if (loading) return <div>loading...</div>;
  if (!team.results) return <div>Not found!</div>;

  return (
    <Container>
      <Header>
        <img src={team.logo} alt={team.name + ' logo'} />
        <h1>{team.name}</h1>
      </Header>
      <Body>
        <StatsContainer>
          <Label>Played:</Label>
          <Stat>{stats.played}</Stat>

          <Label>Points:</Label>
          <Stat>{stats.points}</Stat>

          <Label>Won:</Label>
          <Stat>{stats.won}</Stat>

          <Label>Drawn:</Label>
          <Stat>{stats.drawn}</Stat>

          <Label>Goals for:</Label>
          <Stat>{stats.gf}</Stat>

          <Label>Goals against:</Label>
          <Stat>{stats.ga}</Stat>

          <Label>Goals difference:</Label>
          <Stat>{stats.gd}</Stat>
        </StatsContainer>
        <ResultsContainer>
          <Results results={team.results} />
        </ResultsContainer>
      </Body>
    </Container>
  );
}

export default Team;
