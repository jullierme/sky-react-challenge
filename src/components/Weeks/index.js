import React, { useEffect, useState } from 'react';
import { fetch } from '../../services/api';
import Results from '../Results';
import {
  Container,
  ResultsContainer,
  WeeksContainer,
  Title,
  Week,
} from './styles';

function Weeks({ match }) {
  const [weeks, setWeeks] = useState([]);
  const [chosenWeek, setChosenWeek] = useState([]);
  const [loading, setLoading] = useState(false);

  useEffect(() => {
    fetchWeeks();
  }, []);

  useEffect(() => {
    setChosenWeek(parseInt(match.params.index, 10));
  }, [match.params.index]);

  const fetchWeeks = async () => {
    setLoading(true);

    const data = await fetch(`/weeks`);

    setWeeks(data);

    setLoading(false);
  };

  if (loading) return <div>loading...</div>;

  if (!weeks.length) return <div>Not found!</div>;

  return (
    <Container>
      <Title>Weeks</Title>

      <WeeksContainer>
        {weeks.map((week, index) => (
          <Week
            key={index}
            to={`/weeks/${index + 1}`}
            selected={chosenWeek - 1 === index ? 1 : 0}
          >
            {index + 1}
          </Week>
        ))}
      </WeeksContainer>

      <Title>Results for week #{chosenWeek}</Title>

      <ResultsContainer>
        <Results results={weeks[chosenWeek - 1]} />
      </ResultsContainer>
    </Container>
  );
}

export default Weeks;
