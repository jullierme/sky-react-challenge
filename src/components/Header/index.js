import React from 'react';

import { Container, LinkMenu } from './styles';

function Header() {
  return (
    <Container>
      <LinkMenu to="/table">Table</LinkMenu>
      <LinkMenu to="/weeks/1">Weeks</LinkMenu>
    </Container>
  );
}

export default Header;
