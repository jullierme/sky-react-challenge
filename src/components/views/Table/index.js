import React, { useEffect, useState } from 'react';

import { fetch } from '../../../services/api';
import { getComputedTable } from '../../../services/stats';
import { Link } from 'react-router-dom';

import { TableOfResults, Thead, Tbody, TdTeam } from './styles';

function Table() {
  const [table, setTable] = useState([]);
  const [loading, setLoading] = useState(false);

  useEffect(() => {
    fetchMatches();
  }, []);

  const fetchMatches = async () => {
    setLoading(true);

    const weeksMatches = await fetch(`/weeks`);

    setTable(getComputedTable(weeksMatches));

    setLoading(false);
  };

  const renderTableData = () => {
    return table.map((club, index) => {
      const { id, name, logo, played, won, drawn, gf, ga, gd, points } = club; //destructuring
      const position = index + 1;

      return (
        <tr key={index}>
          <td>{position}</td>
          <TdTeam>
            <Link to={`/teams/${id}`}>
              <img src={logo} alt={name + ' logo'} /> {name}
            </Link>
          </TdTeam>
          <td>{played}</td>
          <td>{won}</td>
          <td>{drawn}</td>
          <td>{gf}</td>
          <td>{ga}</td>
          <td>{gd}</td>
          <td>{points}</td>
        </tr>
      );
    });
  };

  if (loading) return <div>loading...</div>;
  if (!table.length) return <div>Not found!</div>;

  return (
    <TableOfResults>
      <Thead>
        <tr>
          <th>Position</th>
          <th>Club</th>
          <th>Played</th>
          <th>Won</th>
          <th>Drawn</th>
          <th>GF</th>
          <th>GA</th>
          <th>GD</th>
          <th>Points</th>
        </tr>
      </Thead>
      <Tbody>{renderTableData()}</Tbody>
    </TableOfResults>
  );
}

export default Table;
