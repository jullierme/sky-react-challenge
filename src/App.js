import React from 'react';
import { Route, Switch, Redirect } from 'react-router-dom';

import GlobalStyle from './styles/global';

import Header from './components/Header';
import Team from './components/views/Team';
import Weeks from './components/views/Weeks';
import Table from './components/views/Table';

const App = () => (
  <div>
    <Header />
    <Switch>
      <Route path="/weeks/:index" component={Weeks} />
      <Route path="/teams/:index" component={Team} />
      <Route path="/table" component={Table} />
      <Redirect from="/" to="/table" />
    </Switch>
    <GlobalStyle />
  </div>
);

export default App;
