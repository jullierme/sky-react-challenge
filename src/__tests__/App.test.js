import React from 'react';
import { mount, shallow } from 'enzyme';
import { MemoryRouter, Redirect } from 'react-router';

import App from '../App';
import GlobalStyle from '../styles/global';
import Header from '../components/Header';

import Table from '../components/views/Table';
import Weeks from '../components/views/Weeks';
import Team from '../components/views/Team';

describe('App required routers', () => {
  const renderRoutes = path =>
    mount(
      <MemoryRouter initialEntries={[path]}>
        <App />
      </MemoryRouter>
    );

  it('should match snapshot', () => {
    const wrapper = shallow(<App />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render table page when on root', () => {
    const component = renderRoutes('/');

    expect(component.find(Table)).toHaveLength(1);
    expect(component.find(Weeks)).toHaveLength(0);
    expect(component.find(Team)).toHaveLength(0);
  });

  it('should render Team page', () => {
    const component = renderRoutes('/teams/1');

    expect(component.find(Team)).toHaveLength(1);
    expect(component.find(Table)).toHaveLength(0);
    expect(component.find(Weeks)).toHaveLength(0);
  });

  it('should render Weeks page', () => {
    const component = renderRoutes('/weeks/1');

    expect(component.find(Weeks)).toHaveLength(1);
    expect(component.find(Team)).toHaveLength(0);
    expect(component.find(Table)).toHaveLength(0);
  });

  it('should redirect /table when invalid path', () => {
    const component = renderRoutes('/weekssssss');

    expect(component.find(Table)).toHaveLength(1);
  });
});

describe('App required elements', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mount(
      <MemoryRouter>
        <App />
      </MemoryRouter>
    );
  });

  it('should have a GlobalStyle', () => {
    expect(wrapper.find(GlobalStyle).length).toEqual(1);
  });

  it('should have a Header', () => {
    expect(wrapper.find(Header).length).toEqual(1);
  });
});
