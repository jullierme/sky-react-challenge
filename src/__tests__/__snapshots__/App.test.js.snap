// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App required routers should match snapshot 1`] = `
<div>
  <Header />
  <Switch>
    <Route
      component={[Function]}
      path="/weeks/:index"
    />
    <Route
      component={[Function]}
      path="/teams/:index"
    />
    <Route
      component={[Function]}
      path="/table"
    />
    <Redirect
      from="/"
      to="/table"
    />
  </Switch>
  <GlobalStyleComponent />
</div>
`;
